src/fsspec_xrootd/xrootd.py

```diff
@@ -1,46 +1,118 @@
     DirListFlags,
     MkDirFlags,
     OpenFlags,
+    QueryCode,
     StatInfoFlags,
 )
+from XRootD.client.responses import HostList, XRootDStatus  # type: ignore[import]
 
 
 class ErrorCodes(IntEnum):
     INVALID_PATH = 400
 
 
-class XRootDFileSystem(AbstractFileSystem):  # type: ignore[misc]
+def _handle(
+    future: asyncio.Future[tuple[Any, Any]],
+    status: XRootDStatus,
+    content: Any,
+    servers: HostList,
+) -> None:
+    if future.cancelled():
+        return
+    try:
+        future.get_loop().call_soon_threadsafe(future.set_result, (status, content))
+    except Exception as exc:
+        future.get_loop().call_soon_threadsafe(future.set_exception, exc)
+
+
+async def _async_wrap(func: Callable[..., Any], *args: Any) -> Any:
+    future = asyncio.get_running_loop().create_future()
+    status = func(*args, callback=partial(_handle, future))
+    if not status.ok:
+        raise OSError(status.message.strip())
+    return await future
+
+
+def _make_vectors(
+    file_ranges: list[tuple[int, int]],
+    max_num_chunks: int,
+    max_chunk_size: int,
+) -> list[list[tuple[int, int]]]:
+    modified_file_ranges = []
+    for fr in file_ranges:
+        r = fr[1] - fr[0]
+        if r > max_chunk_size:
+            a, b = divmod(r, max_chunk_size)
+            for j in range(0, a):
+                modified_file_ranges.append(
+                    (
+                        fr[0] + max_chunk_size * j,
+                        fr[0] + (j + 1) * max_chunk_size,
+                    )
+                )
+            modified_file_ranges.append((fr[0] + max_chunk_size * a, fr[1]))
+        else:
+            modified_file_ranges.append(fr)
+    result: list[list[tuple[int, int]]] = []
+    for i in range(0, len(modified_file_ranges)):
+        if i % max_num_chunks == 0:
+            result.append([])
+        result[-1].append(modified_file_ranges[i])
+
+    vectors: list[list[tuple[int, int]]] = []
+    for vec in result:
+        vectors.append([])
+        for v in vec:
+            vectors[-1].append((v[0], v[1] - v[0]))
+    return vectors
+
+
+class XRootDFileSystem(AsyncFileSystem):  # type: ignore[misc]
 
     protocol = "root"
     root_marker = "/"
     default_timeout = 60
+    async_impl = True
+
+    cat_chunk_cache: dict[str, Any] = defaultdict(dict)
 
-    def __init__(self, *args: list[Any], **storage_options: Any) -> None:
+    def __init__(
+        self,
+        *args: list[Any],
+        asynchronous: bool = False,
+        loop: Any = None,
+        batch_size: int | None = None,
+        **storage_options: Any,
+    ) -> None:
         self.timeout = storage_options.get("timeout", XRootDFileSystem.default_timeout)
         self._myclient = client.FileSystem(
             XRootDFileSystem.protocol + "://" + storage_options["hostid"]
         )
         status, _n = self._myclient.ping(15)
         if not status.ok:
             raise OSError(f"Could not connect to server {storage_options['hostid']}")
```

Comment:
wait, now it doesnt pass the test_broken_server() test

Reply:
I think the `test_broken_server` test will need to be modified to actually try to open a file, at which point it should fail

Reply:
I modified the test so it actually tries to read from the file and it fails as expected. It does take 60 seconds to do the test because thats the default timeout